### DISCUSSION TITLE
Opening files on macOS

### DISCUSSION THREAD

**Discussion by arklumpus** [AvaloniaUI/Avalonia#5504] [Q&A] [answered]
Is there a recommended way to make it possible to open files (by double-click or drag-n-drop on the app icon) on macOS?

On Windows and Linux, my program takes input files as command-line arguments, checks if another instance is running and then decides: if another instance is already running, it sends the file names through a named pipe and exits; otherwise it initialises the whole Avalonia framework and opens the files.

This _would_ work on macOS as well (i.e. it works if you run the program manually from the command line), except that when you open a file "with" an app on macOS, macOS does not provide the file name as a command line argument.

The workaround in #3822 might be a starting point, but that seems to be specific to opening URLs rather than files, and I have a hard time navigating the Apple documentation website to find out what should be used instead of `handleGetURLEvent:withReplyEvent:` to subscribe to the correct event...

My current solution is to use an Automator app that receives the files as input and sends them as command line arguments to a shell script; the shell script then runs my app using `open -n Application.app --args "$@"`. This works, but has a few annoying drawbacks:

* You get two icons in the dock, one for the Avalonia app that is actually running, and one for the Automator app. Neither one can be removed, because removing the Avalonia app would mean that the user cannot meta-tab into it either, and removing the Automator app icon would prevent drag-n-drop on the dock icon (because the Avalonia app itself does not accept files as input).

* The app needs to have full disk access permissions. This seems to be because using `open` breaks parent-child relationships and thus, even though the user has given permission for the Automator app to access the file, this is not inherited by my Avalonia app.

Is there a better way to achieve this?

**Comment by mtrellet** [comment]
Unfortunately I cannot provide a solution to the main question here, on the contrary, we would be really interested to know whether there are some best practices existing when it comes to open a file associated to an Avalonia app. 

We encounter the same issue and our quick observation (and issue) at this point is that no path is added to the `Args` array of the `IClassicDesktopStyleApplicationLifetime` object upon the application start. So of course no way to tell the application which file(s) to open.

Thanks in advance for any clue about it!

**Reply by arklumpus** [reply]
From my investigations, I think the problem is that adding files to the command line arguments is not the "macOS way" of dealing with file associations. Mac apps apparently subscribe to some sort of an "event" that is fired when the user double-clicks a file or drags it onto the app's icon on the dock (if the app is not open when this happens, it is started first). We need essentially a way to hook into this event and get the file path from there (in theory this should automatically give the app permission to open the file without it requiring full disk access).

I think a reasonable way this could be implemented would require two things:
1. A parameter "somewhere" in Avalonia that tells us whether the app was started normally or by opening one or more files (and, in this case, the paths to these files) - e.g. a `string[]` which is `null` or has length 0 when the app is started normally and is populated when it is started from one or more files.
2. An event in Avalonia that is triggered when the macOS file opened event fires after the app is already open; here the file path(s) would be contained in the `EventArgs`.

Internally Avalonia would somehow hook to the macOS system event and then trigger the relevant response.

Of course, I have no idea on how to implement this 😅 (because I am not very familiar with the Avalonia source nor with macOS app development)... I also think things may be even more complicated because  macOS apps are somehow disconnected from the "traditional" concept of an executable.

In the mean time, if you need it, I think my solution of having the Avalonia app nested within an Automator app that forwards the file names (you can see an implementation [here](https://github.com/arklumpus/TreeViewer/tree/main/src/Resources/Mac-x64)) is a reasonable approximation, though with the two caveats above... Happy to provide more details if you need them!

**Reply by jp2masa** [reply]
Not sure if it's the same thing, but URL handling on macOS was implemented in #5639. When URLs are opened, the [`Application.UrlsOpened`](https://github.com/AvaloniaUI/Avalonia/blob/afb8ae6f3c517dae912729511483995b16cb31af/src/Avalonia.Controls/Application.cs#L59) event is invoked.

**Reply by mtrellet** [reply]
@arklumpus Thanks a lot for the detailed explanation and the leads to solve the issue. Your reasoning makes a lot of sense and that would be nice to have that implemented in a next release of Avalonia indeed. But unfortunately I'm like you, with very limited experience on Avalonia core code and OSX app development...

@jp2masa We'll give it a quick try to see if by any chance the same event is triggered upon file opening and if the file paths could be obtained from there. I must admit that I do not have high hopes but this is worth checking, thanks for the tip!

**Comment by arklumpus** [comment, marked as answer]
I finally had time to test what @jp2masa suggested, and I can confirm that `Application.UrlsOpened` actually works!

When you create a proper app bundle and associate it with a file extension (through the `Info.plist` file), then:

* If the app is running, double-clicking on a file causes the `Application.UrlsOpened` event to fire, and the `Urls` property of the `UrlOpenedEventArgs` contains a URL (with the `file://` protocol) with the full path to the file.
* If the app is not running, double-clicking on a file launches the app and then fires the event (*).
* If one or more files are dragged on the app's icon in the dock, the `Urls` contain one element for each file.

(*) **Note**: from my experiments, there appears to be a limited time frame to subscribe to the event, if you wish to catch the event that is fired when the app is opened by clicking on an associated file. A delay of over 300ms seems to miss the event entirely, while a delay over 250ms sometimes works and sometimes doesn't; with 250ms or less, it always seems to work. These values might also be dependant on the machine (e.g. faster computers might have a different tolerance than slower computers); I tested them with a late 2014 Mac mini with a 1.4GHz i5 processor and 4GB of RAM.

I suppose the best thing to do would be to subscribe to the event as soon as possible (e.g. in the `App`'s `OnFrameworkInitializationCompleted` method?). If the program needs some more time to be able to properly handle the event, I would do something like this:

1. Create two `System.Threading.SemaphoreSlim` instances:
    ```CSharp
    SemaphoreSlim initialisationSemaphore = new SemaphoreSlim(0, 1);
    SemaphoreSlim eventSemaphore = new SemaphoreSlim(1, 1);
    ```

2. In the event handler, asynchronously wait on both:
    ```CSharp
    this.UrlsOpened += async (s, e) =>
    {
        await eventSemaphore.WaitAsync();
        await initialisationSemaphore.WaitAsync();

        // Do something with e.Urls

        initialisationSemaphore.Release();
        eventSemaphore.Release();
    };
    ```
    Assuming this is within the `App` class - otherwise, replace `this` with `Application.Current`.

3. After the app has been completely initialised and is ready to handle the event, call `initialisationSemaphore.Release()`.

With this approach, when the `UrlsOpened` event first fires (i.e. just after the app is launched by opening the file), the handle will immediately acquire the `eventSemaphore` (because it's already available), and then wait for the `initialisationSemaphore`. When the app has been initialised and the `initialisationSemaphore` is released, the handler proceeds to actually handle the event. On subsequent events, the handler proceeds immediately because the semaphores are already available.

If the user opens another file while the app is being initialised (and thus, the handler is still waiting for the `initialisationSemaphore`), the second handler (if it even fires, see below) will wait for the `eventSemaphore` before proceeding, which will be released once the first handler has finished its thing. This also works if the app was opened directly, instead of double-clicking a file.

Caveate: I haven't tested this code yet; the `eventSemaphore` is _probably_ overkill, if we can assume that there won't be two events firing at the same time (because the UI is single-threaded), but I think using it is safer and shouldn't introduce any noticeable overhead.

**Comment by maxkatz6** [comment]
This question was answered in 2021. Now, there is another and documented API available: https://docs.avaloniaui.net/docs/concepts/services/activatable-lifetime
